From 78c25fcf97b2f84df8eabbc8631be3e936c4a542 Mon Sep 17 00:00:00 2001
From: David Morel <david.morel@vates.tech>
Date: Tue, 11 Jun 2024 08:39:26 +0200
Subject: [PATCH] ovsdb-server: custom openssl config to limit ciphers and
 protocols

- Add an XCP-ng specific openssl configuration file limiting the ciphers
  and protocols used by ovsdb-server, to avoid weak ciphers
- Add an OPENSSL_CONF environment variable to point to the new
  configuration file at startup

Signed-off-by: David Morel <david.morel@vates.tech>
---
 .../etc_openvswitch_openssl-xcpng-ovsdb.conf  | 20 +++++++++++++++++++
 ...usr_lib_systemd_system_openvswitch.service |  1 +
 2 files changed, 21 insertions(+)
 create mode 100644 xcp-ng/etc_openvswitch_openssl-xcpng-ovsdb.conf

diff --git a/xcp-ng/etc_openvswitch_openssl-xcpng-ovsdb.conf b/xcp-ng/etc_openvswitch_openssl-xcpng-ovsdb.conf
new file mode 100644
index 0000000..b2184c1
--- /dev/null
+++ b/xcp-ng/etc_openvswitch_openssl-xcpng-ovsdb.conf
@@ -0,0 +1,20 @@
+# XCP-ng's OpenSSL configuration file to restrict TLS versions to TLS 1.2 and
+# TLS 1.3 so ovsdb-server won't expose older protocols. Requires an openvswitch
+# version linked with xs-openssl as openssl 1.0.2 has no support for such a
+# configuration.
+
+openssl_conf = openssl_init
+
+[openssl_init]
+ssl_conf = ssl_sect
+
+[ssl_sect]
+system_default = system_default_sect
+
+[system_default_sect]
+MinProtocol = TLSv1.2
+MaxProtocol = TLSv1.3
+
+# Althrough xs-openssl should now be built with no-weak-ciphers, specify a list
+# here, with a default value, but we could easily make a custom list if needed.
+CipherString = DEFAULT@SECLEVEL=2
diff --git a/xenserver/usr_lib_systemd_system_openvswitch.service b/xenserver/usr_lib_systemd_system_openvswitch.service
index 31f18ef..0e071bc 100644
--- a/xenserver/usr_lib_systemd_system_openvswitch.service
+++ b/xenserver/usr_lib_systemd_system_openvswitch.service
@@ -7,6 +7,7 @@ PartOf=network.target
 [Service]
 Type=forking
 Environment=ASAN_OPTIONS=abort_on_error=1:disable_coredump=0:unmap_shadow_on_exit=1
+Environment=OPENSSL_CONF=/etc/openvswitch/openssl-xcpng-ovsdb.conf
 ExecStart=/usr/share/openvswitch/scripts/ovs-start
 ExecStop=/usr/share/openvswitch/scripts/ovs-ctl stop
 OOMScoreAdjust=-1000
-- 
2.45.1

